cunumeric/array.py

```diff
@@ -1741,6 +1741,63 @@ def astype(
         result._thunk.convert(self._thunk, warn=False)
         return result
 
+    def take(self, indices, axis=None, out=None, mode="raise"):
+        if not np.isscalar(indices):
+            # if indices is a tuple or list, bring sub-tuples to the same shape
+            # and concatenate them
+            indices = convert_to_cunumeric_ndarray(indices)
+
+        if axis is None:
+            self = self.ravel()
+            axis = 0
+        elif axis < 0:
+            axis = self.ndim + axis
+        elif axis >= self.ndim:
+            raise ValueError("axis is out of bounds for array of dimension")
+
+        if mode == "raise":
+            if np.isscalar(indices):
+                if (indices < 0) or (indices >= self.shape[axis]):
+                    raise ValueError("invalid entry in indices array")
+            else:
+                if (indices < 0).any() or (indices >= self.shape[axis]).any():
+                    raise ValueError("invalid entry in indices array")
+        elif mode == "wrap":
+            indices = indices % self.shape[axis]
+        elif mode == "clip":
+            if np.isscalar(indices):
+                if indices >= self.shape[axis]:
+                    indices = self.shape[axis] - 1
+                if indices < 0:
+                    indices = 0
+            else:
+                indices = indices.clip(0, self.shape[axis] - 1)
+        else:
+            raise ValueError(
+                "Invalid mode '{}' for take operation".format(mode)
+            )
+        if self.shape[axis] == 0:
+            if indices.size != 0:
+                raise IndexError(
+                    "Cannot do a non-empty take() from an empty axis."
+                )
+            return self
+
+        point_indices = tuple(slice(None) for i in range(0, axis))
+        point_indices += (indices,)
+        res = self[point_indices]
+        if out is not None:
+            out = convert_to_cunumeric_ndarray(out)
+            if out.shape != res.shape:
+                raise ValueError("Shape mismatch: out array has wrong shape")
+            if out.dtype != res.dtype:
+                raise ValueError("Type mismatch: out array has wrong type")
+
+            out._thunk = res._thunk
```

Comment:
In most cases we cannot set the `._thunk` field to something else after it's been initialized, otherwise views won't work correctly (we instead update the contents of the thunk, by going through the methods in `eager.py` and `deferred.py`). In this case, if the `out` array had a view pointing to it before entry to this function, then that view would not be updated, because you are essentially creating a new region and setting `out` to it, instead of updating the contents of the region that was backing the original `out`:

```
a = cn.zeros((3,))
b = a[:]
c = cn.ones((2,3))
c.take(0, axis=0, out=a)
assert np.array_equal(a, b)
```

Reply:
should I replace it with `out._thunk.copy(res._thunk)` ?

Reply:
That creates an unnecessary copy, I think you could just do `out[:] = self[point_indices]`.

Reply:
In this case I will perform advanced indexing operation again (which includes copy). Wouldn't it be better to call `copy`? I need to keep `res = self[point_indices]` above to check `out.shape == res.shape`.

Reply:
I don't think you need to create `res` if `out is not None`. `out[:] = self[point_indices]` will fail if the shape is incompatible, and you can use `self.dtype` in place of `res.dtype`.